### PR TITLE
Addresses the problem of an HA instance being partitioned away from the cluster

### PR DESCRIPTION
**Pull request by digitalstain**
This is a set of changes that make explicit the behaviour of instances or groups of instances that are no longer in contact with a majority of the HA cluster. The various commits introduce the ability to detect that state, act on it and revert to read only mode and help with the handling of some edge cases.

changelog: [2.3, HA] Partitioned instances revert to PENDING state while allowing read only operations

### CODE REVIEW

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterPartitionIT.java [hunk 1]
```diff
@@ -144,6 +142,7 @@ public void losingQuorumIncrementallyShouldMakeAllInstancesPendingAndReadOnly()
                 .withProvider( ClusterManager.clusterOfSize( clusterSize ) )
                 .withSharedConfig( stringMap(
                         ClusterSettings.heartbeat_interval.name(), "1",
+                        ClusterSettings.heartbeat_timeout.name(), "3",
```
Comment (MishaDemianenko) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterPartitionIT.java]: maybe it's too low for aws machines

Comment (digitalstain) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterPartitionIT.java]: Unfortunately that may be true. I'll amend.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberState.java [hunk 2]
```diff
@@ -346,4 +378,23 @@ public abstract HighAvailabilityMemberState masterIsAvailable( HighAvailabilityM
     public abstract boolean isEligibleForElection();
 
     public abstract boolean isAccessAllowed();
+
+    private boolean transitionWasLegal = true;
```
Comment (MishaDemianenko) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberState.java]: please move those fields to be before any methods, just after SLAVE enum value declaration finish

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberStateMachine.java [hunk 3]
```diff
@@ -323,5 +357,32 @@ private long getTotalCount()
         {
             return Iterables.count( members.getMembers() );
         }
+
+        private boolean checkNewStateAndTransition( HighAvailabilityMemberState newState )
```
Comment (MishaDemianenko) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberStateMachine.java]: i think we need to spilt it into 2 separate methods and keep assignment of new state outside of this method

Comment (MishaDemianenko) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberStateMachine.java]: @digitalstain still previous comment, wdyt?

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberStateMachine.java [hunk 4]
```diff
@@ -348,5 +357,27 @@ private long getTotalCount()
...
+                log.warn( format( "Message received resulted in illegal state transition. I was in state %s, " +
+                        "context was %s. The error message is %s. This instance will now transition to PENDING state " +
+                        "and " +
+                        "ask for new elections. While this may fix the error, it may indicate that there is some " +
+                        "connectivity issue or some instability of cluster members.", state, context, newState
+                        .errorMessage() ) );
+                context.setElectedMasterId( null );
+                context.setAvailableHaMasterId( null );
+                changeStateToPending();
+                election.performRoleElections();
```
Comment (MishaDemianenko) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberStateMachine.java:379]: looks like `return false` is gone? hmmm

**Comment by MishaDemianenko** [comment]
@digitalstain another round of review done, couple of question/comments
